### ISSUE TITLE
Enable GSSAPI authentication support when using git

### ISSUE THREAD

**Issue by hktonylee** [rust-lang/cargo#6295]
**Problem**
Although I can use the `git clone ...` in my command line normally, the`cargo run` reports authentication error. The detailed log is here:

```
> cargo build  --verbose
    Updating git repository `ssh://some.internal.host/pkg/SomePackage`
error: failed to load source for a dependency on `some_package`

Caused by:
  Unable to update ssh://some.internal.host/pkg/SomePackage

Caused by:
  failed to fetch into /Users/userA/.cargo/git/db/SomePackage-0000000000000000

Caused by:
  failed to authenticate when downloading repository
attempted ssh-agent authentication, but none of the usernames `userA`, `git` succeeded

Caused by:
  no authentication available
```

**Steps**

1. Added `some_package = { git = "ssh://some.internal.host/SomePackage" }` to `Cargo.toml`
2. Ran `cargo run` in command line
3. It reported the above message

**Possible Solution(s)**

The first solution is probably to support GSSAPI authentication. I don't know if this is related but my company uses Kerberos as the authentication agent. Here is the output when I run `env GIT_SSH_COMMAND="ssh -vvv" git clone ssh://....`:

```
...
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic,keyboard-interactive
debug3: start over, passed a different list publickey,gssapi-keyex,gssapi-with-mic,keyboard-interactive
debug3: preferred gssapi-with-mic,publickey,keyboard-interactive,password
debug3: authmethod_lookup gssapi-with-mic
debug3: remaining preferred: publickey,keyboard-interactive,password
debug3: authmethod_is_enabled gssapi-with-mic
debug1: Next authentication method: gssapi-with-mic
debug3: send packet: type 50
debug2: we sent a gssapi-with-mic packet, wait for reply
...
debug1: Authentication succeeded (gssapi-with-mic).
...
```

As you can see the system ssh can authenticate with gssapi-with-mic without problems. So my guess is that the Cargo does not use the system ssh, or it does not enable GSSAPI support. But this is only my guess and may be wrong =)

The second solution is to add more verbose output when downloading from remote git (especially ssh which is a beast). So that I can debug myself.

**Notes**

Output of `cargo version`:
cargo 1.29.0 (524a578d7 2018-08-05)

macOS version 10.12.6
git version 2.18.0

**Comment by alexcrichton** [comment]
Thanks for the report! Support for this may be unlikely to be added to libgit2 (the default library), but we've recently added the [`net.git-fetch-with-cli`](https://doc.rust-lang.org/nightly/cargo/reference/config.html) configuration option to solve cases like this. Want to give that a spin and see if it works for you?

**Comment by hktonylee** [comment]
Thank you Alex 👍  Your suggestion works :D

**Comment by TitanThinktank** [comment]
i am gettign that error for Rocket framework and no matter what i do its not going away, now i wait patiently for someone to show the solution, i can wait for days but then i will give up.

**Comment by hktonylee** [comment]
I think Alex's solution works well. But that may be a little bit implicit because I need to check several reference to get the answer. Let me write down the step-by-step guideline here

1. Create a file at `<project dir>/.cargo/config` as shown in  [Hierarchical structure](https://doc.rust-lang.org/nightly/cargo/reference/config.html#hierarchical-structure)
1. Modify the file, add these lines to it
    ```
    [net]
    git-fetch-with-cli = true
    ```
1. `rustup` the nightly build
    ```bash
    rustup update # optional
    rustup install nightly
    rustup default nightly
    ```
1. Remove the folder `<project dir>/target `
1. Run `cargo build` and see if it can build

Sorry that I did not have time to verify the steps. I tried my best to recall them.
